File: src/Coachable/4-DesignDocs/DD_8.py
"""
463. Island Perimeter
https://leetcode.com/problems/island-perimeter/description/
"""

from typing import List


class Solution:
    def islandPerimeter(self, grid: List[List[int]]) -> int:

        def dfs(r: int, c: int) -> int:
            if r < 0 or c < 0 or r > len(grid) or c > len(grid[0]):
                return 1
            if (r, c) in visited:
                return 0

            visited.add((r, c))

            return dfs(r - 1, c) + dfs(r + 1, c) + dfs(r, c - 1) + dfs(r, c + 1)

        visited = set()

        for row_idx in range(len(grid)):
            for col_idx in range(len(grid[0])):
                if grid[row_idx][col_idx] == 1:
                    return dfs(row_idx, col_idx)

File: src/misc/LearningMaterial/4-DesignDocs/DD_8.rs
/*
463. Island Perimeter
https://leetcode.com/problems/island-perimeter/description/
*/

use std::collections::HashSet;

fn main() {
    // Example usage:
    let grid = vec![
        vec![0, 1, 0, 0],
        vec![1, 1, 1, 0],
        vec![0, 1, 0, 0],
        vec![1, 1, 0, 0],
    ];
    let result = island_perimeter(grid);
    println!("Island perimeter: {}", result);
}

fn island_perimeter(grid: Vec<Vec<i32>>) -> i32 {
    let mut visited: HashSet<(i32, i32)> = HashSet::new();

    fn dfs(grid: &Vec<Vec<i32>>, visited: &mut HashSet<(i32, i32)>, r: usize, c: usize) -> i32 {
        if r >= grid.len() || c >= grid[0].len() || grid[r][c] == 0 {
            return 1;
        }
        if visited.contains(&(r as i32, c as i32)) {
            return 0;
        }

        visited.insert((r as i32, c as i32));

        dfs(grid, visited, r.wrapping_sub(1), c)
            + dfs(grid, visited, r + 1, c)
            + dfs(grid, visited, r, c.wrapping_sub(1))
            + dfs(grid, visited, r, c + 1)
    }

    for r in 0..grid.len() {
        for c in 0..grid[0].len() {
            if grid[r][c] == 1 {
                return dfs(&grid, &mut visited, r, c);
            }
        }
    }
    0
}
